app/api/items/fetch_active_attempt.go

```diff
@@ -0,0 +1,136 @@
+package items
+
+import (
+	"errors"
+	"net/http"
+	"strconv"
+
+	"github.com/go-chi/render"
+	"github.com/jinzhu/gorm"
+
+	"github.com/France-ioi/AlgoreaBackend/app/database"
+	"github.com/France-ioi/AlgoreaBackend/app/formdata"
+	"github.com/France-ioi/AlgoreaBackend/app/service"
+	"github.com/France-ioi/AlgoreaBackend/app/token"
+)
+
+func (srv *Service) fetchActiveAttempt(w http.ResponseWriter, r *http.Request) service.APIError {
+	var err error
+
+	itemID, err := service.ResolveURLQueryPathInt64Field(r, "item_id")
+	if err != nil {
+		return service.ErrInvalidRequest(err)
+	}
+
+	user := srv.GetUser(r)
+	if err = user.Load(); err == database.ErrUserNotFound {
+		return service.InsufficientAccessRightsError
+	}
+	service.MustNotBeError(err)
+
+	var itemInfo struct {
+		HasAttempts       bool    `gorm:"column:bHasAttempts"`
+		AccessSolutions   bool    `gorm:"column:accessSolutions"`
+		HintsAllowed      bool    `gorm:"column:bHintsAllowed"`
+		TextID            *string `gorm:"column:sTextId"`
+		URL               string  `gorm:"column:sUrl"`
+		SupportedLangProg *string `gorm:"column:sSupportedLangProg"`
+	}
+	err = srv.Store.Items().Visible(user).Where("ID = ?", itemID).
+		Where("partialAccess > 0 OR fullAccess > 0").
+		Where("items.sType IN('Task','Course')").
+		Select("accessSolutions, bHasAttempts, bHintsAllowed, sTextId, sUrl, sSupportedLangProg").
+		Take(&itemInfo).Error()
+	if gorm.IsRecordNotFoundError(err) {
+		return service.InsufficientAccessRightsError
+	}
+	service.MustNotBeError(err)
+
+	var userItemInfo struct {
+		ActiveAttemptID  *int64  `gorm:"column:idAttemptActive"`
+		HintsRequested   *string `gorm:"column:sHintsRequested"`
+		HintsCachedCount int32   `gorm:"column:nbHintsCached"`
+	}
+	apiError := service.NoError
+	err = srv.Store.InTransaction(func(store *database.DataStore) error {
+		userItemStore := store.UserItems()
+		service.MustNotBeError(userItemStore.CreateIfMissing(user.UserID, itemID))
+		service.MustNotBeError(userItemStore.Where("idUser = ?", user.UserID).Where("idItem = ?", itemID).
```

Comment:
You could "left join" it with the previous request to prevent another request (even 2 requests?) in case it already exists. I don't know if it worth the change.

Reply:
There is a call to 'CreateIfMissing' between the queries. I think, it forces us to use separate requests. Anyway, this function is complex enough and JOINs would make it even more complex, I think.